### ISSUE TITLE
Can not delivery pushes to test Android devices in China

### ISSUE THREAD

**Issue by Gelassen** [ParsePlatform/Parse-SDK-Android#236]
Working on integration of parse platform in our application. We had successful experience in the past. 

However at this moment by some reasons notifications don't come to Android devices in China. For Russia region it works. 

We verified device is registered and listed in dashboard. 

Please note that deviceId is empty on dashboard (push type GCM or "undefined") and within app accessed as 

        parseInstallation.saveInBackground(new SaveCallback() {
            @Override
            public void done(ParseException e) {
                Log.e(App.PARSE, e == null ? "Succeed to register the app" : "Failed to register the app", e);
                String deviceToken = (String) ParseInstallation.getCurrentInstallation().get("deviceToken");
                Log.d(App.PARSE, "Device token: " + deviceToken);
            }
        });

For devices without GCM app is also registered and delivery report notice that push was sent for this devices. 

To summarise. 
China region: device with GCM is sucessfuly registered, doesn't send the device token and report noticed push has not delivered to this device. Notification is not appeared
China region: device without GCM is sucessfuly registered, push is sent to them based on report. Notification is not appeared 

Non-China region: device with GCM is sucessfuly registered, device token is existed. Notification is appeared
Non-China region: device without GCM is sucessfuly registered, push is sent to them based on report. Notification is appeared

**Comment by Gelassen** [comment]
Update: 
If I use VPN for China region I am able to register app with device token and even receive push notifications.

**Comment by grantland** [comment]
Could you clarify if you're still having an issue or that it's resolved? If it's not resolved, could you elaborate on the different environments you're using to test?

**Comment by Gelassen** [comment]
Hello, 

Issue is reproduced, still working on it. 

We have distributed team between Russian and Chinese offices. Both teams have smartphones with GCM, e.g. Nexus 5, and without it, e.g. Xiaomi Mi Note. 

An all devices in Russia push notifications works, on all devices in China pushes don't work, except case when it was run via VPN. 

Please let me know if you have others questions. I will be glad to help.

**Comment by Gelassen** [comment]
Decided to try default demo project -- Parse-Starter-Project -- to exclude all possible 3rd party dependencies. The same issue, it doesn't work in China region.

**Comment by grantland** [comment]
As far as I'm aware, Google Play Services do not work in China, which Parse's GCM implementation relies on. Are these devices able to connect to Google Play and other Google services without VPN in China?

**Comment by Gelassen** [comment]
"As far as I'm aware, Google Play Services do not work in China, which Parse's GCM implementation relies on. Are these devices able to connect to Google Play and other Google services without VPN in China?"
--
It is correct. However parse.com claims it work in China region. 
https://parse.com/questions/can-your-push-function-work-in-china
https://parse.com/questions/use-parse-in-china
https://parse.com/questions/will-parse-notification-works-on-china-where-google-services-are-blocked

Moreover, we had the project which successfully used Parse for notifications -- it was android hybrid app (js\java).

Both Nexus 5 and Xiaomi MI Note doesn't work in China even with test of sample app.

**Comment by grantland** [comment]
Unfortunately these answers are over a year old and the Chinese government's firewall is susceptible to change, so our services may be blocked in China at any given time.

However, we haven't had any other reports of this happening, so I want to figure out if this is a firewall issue or configuration issue. Could add the following information:

* The SDK version you're using
* Enable logging with `Parse.setLogLevel(Parse.LOG_LEVEL_VERBOSE);` and paste the output from your applications run. Specifically I'm looking for the output of this statement: https://github.com/ParsePlatform/Parse-SDK-Android/blob/master/Parse/src/main/java/com/parse/ManifestInfo.java#L266
* Your `build.gradle` dependencies
* Any `jar`s you have in your project
* The `pushType` of the devices with and without GCM in your dashboard

**Comment by Gelassen** [comment]
The SDK version:
1.10.3

Dependencies:
dependencies {
    compile 'com.android.support:appcompat-v7:22.2.0'

    compile 'com.parse.bolts:bolts-android:1.2.1'
    compile 'com.parse:parse-android:1.10.3'
}

Logs:
11-10 11:36:48.092  14755-14755/? I/art﹕ Late-enabling -Xcheck:jni
11-10 11:36:48.182  14755-14755/com.parse.starter V/com.parse.ParseHttpClient﹕ Using net.java.URLConnection library for networking communication.
11-10 11:36:48.210  14755-14755/com.parse.starter V/com.parse.ManifestInfo﹕ Using gcm for push.
11-10 11:36:48.266  14755-14777/com.parse.starter V/com.parse.CachedCurrentInstallationController﹕ Successfully deserialized Installation object
11-10 11:36:48.276  14755-14778/com.parse.starter V/com.parse.GcmRegistrar﹕ Sending GCM registration intent
11-10 11:36:48.344  14755-14795/com.parse.starter D/OpenGLRenderer﹕ Use EGL_SWAP_BEHAVIOR_PRESERVED: true
11-10 11:36:48.352  14755-14755/com.parse.starter D/Atlas﹕ Validating map...
11-10 11:36:48.392  14755-14795/com.parse.starter I/Adreno-EGL﹕ <qeglDrvAPI_eglInitialize:379>: QUALCOMM Build: 01/14/15, ab0075f, Id3510ff6dc
11-10 11:36:48.393  14755-14795/com.parse.starter I/OpenGLRenderer﹕ Initialized EGL, version 1.4
11-10 11:36:48.418  14755-14795/com.parse.starter D/OpenGLRenderer﹕ Enabling debug mode 0

External libs:
< Android API 22 Platform >
< jdk 1.7 >
appcompat-v7-22.2.0
bolts-android-1.2.1
parse-android-1.10.3
support-annotations-22.2.0
support-v4-22.2.0

Push type:
GCM and "undefined"

**Comment by Gelassen** [comment]
Hello, any news so far?

**Comment by grantland** [comment]
Is this an existing project that has worked in China before and doesn't now, or is this a completely new project?

**Comment by wangmengyan95** [comment]
Hi @Gelassen, it seems your test device has GCM, so we use GCM to send notifications (`11-10 11:36:48.210 14755-14755/com.parse.starter V/com.parse.ManifestInfo﹕ Using gcm for push`). Since you mentioned there is no deviceToken, I think the case is the device tries to talk to GCM for the deviceToken(`11-10 11:36:48.276 14755-14778/com.parse.starter V/com.parse.GcmRegistrar﹕ Sending GCM registration intent`), but since GCM is blocked in China, the device can not get the response. We use deviceToken to identify devices so without it we can not deliver push notifications. There is nothing more we can do, it is a problem of China Firewall.
Could you also paste the log for the device which pushType is not GCM?

**Comment by Gelassen** [comment]
@grantland, it is new project, we also have another one with parse integration it worked well a half year ago.

@wangmengyan95, as I shown above parse promise to work in China region. Docs doesn't mention any limits to devices with\without GCM, moreover if you have implemented solution to work inside of China (it means without GCM\Google play) GCM availability should not be an issue. 
I understand China's investments in their firewall has an impact on availability of all web resources and network communication. The question is does parse are going to work on it or we can expect post in the nearest future that parse doesn't work in China anymore?  

Below is log for device without GCM. By some reason it is different from 3 days ago, it failed to register app via PPNS. Also my team found the app failed to find PPNS.jar if we use gradle dependency. We add it manually ti lib folder to resolve this issue, 
 
11-12 10:20:26.729  26983-26983/com.parse.starter D/dalvikvm﹕ Late-enabling CheckJNI
11-12 10:20:26.779  26983-26983/com.parse.starter I/MultiDex﹕ VM with version 1.6.0 does not have multidex support
11-12 10:20:26.779  26983-26983/com.parse.starter I/MultiDex﹕ install
11-12 10:20:26.789  26983-26983/com.parse.starter I/MultiDex﹕ MultiDexExtractor.load(/data/app/com.parse.starter-1.apk, false)
11-12 10:20:26.789  26983-26983/com.parse.starter I/MultiDex﹕ Detected that extraction must be performed.
11-12 10:20:26.799  26983-26983/com.parse.starter I/MultiDex﹕ load found 0 secondary dex files
11-12 10:20:26.799  26983-26983/com.parse.starter I/MultiDex﹕ install done
11-12 10:20:26.819  26983-26983/com.parse.starter V/com.parse.ParseHttpClient﹕ Using net.java.URLConnection library for networking communication.
11-12 10:20:26.869  26983-26983/com.parse.starter V/com.parse.ManifestInfo﹕ Using ppns for push.
11-12 10:20:26.899  26983-27011/com.parse.starter V/com.parse.CachedCurrentInstallationController﹕ Successfully deserialized Installation object
11-12 10:20:26.929  26983-26983/com.parse.starter D/dalvikvm﹕ GetMethodID: not returning static method Landroid/os/Process;.getTotalMemory ()J
11-12 10:20:26.929  26983-26983/com.parse.starter D/dalvikvm﹕ GetMethodID: not returning static method Landroid/os/Process;.getFreeMemory ()J
11-12 10:20:26.949  26983-26983/com.parse.starter I/dalvikvm﹕ Could not find method android.content.res.TypedArray.getChangingConfigurations, referenced from method android.support.v7.internal.widget.TintTypedArray.getChangingConfigurations
11-12 10:20:26.949  26983-26983/com.parse.starter W/dalvikvm﹕ VFY: unable to resolve virtual method 454: Landroid/content/res/TypedArray;.getChangingConfigurations ()I
11-12 10:20:26.949  26983-26983/com.parse.starter D/dalvikvm﹕ VFY: replacing opcode 0x6e at 0x0002
11-12 10:20:26.959  26983-26983/com.parse.starter I/dalvikvm﹕ Could not find method android.content.res.TypedArray.getType, referenced from method android.support.v7.internal.widget.TintTypedArray.getType
11-12 10:20:26.959  26983-26983/com.parse.starter W/dalvikvm﹕ VFY: unable to resolve virtual method 476: Landroid/content/res/TypedArray;.getType (I)I
11-12 10:20:26.959  26983-26983/com.parse.starter D/dalvikvm﹕ VFY: replacing opcode 0x6e at 0x0002
11-12 10:20:27.019  26983-26983/com.parse.starter V/PPNSService﹕ Received request to start service if required
11-12 10:20:27.029  26983-27024/com.parse.starter V/com.parse.PushConnection﹕ PushConnection transitioning from com.parse.PushConnection$WaitStartState@4222cdc8 to com.parse.PushConnection$ConnectState@4222e508
11-12 10:20:27.049  26983-26983/com.parse.starter I/Adreno-EGL﹕ <qeglDrvAPI_eglInitialize:410>: EGL 1.4 QUALCOMM build:  ()
    OpenGL ES Shader Compiler Version: E031.24.00.15
    Build Date: 04/21/15 Tue
    Local Branch:
    Remote Branch:
    Local Patches:
    Reconstruct Branch:
11-12 10:20:27.089  26983-26983/com.parse.starter D/OpenGLRenderer﹕ Enabling debug mode 0
11-12 10:20:27.189  26983-26983/com.parse.starter I/Timeline﹕ Timeline: Activity_idle id: android.os.BinderProxy@420b80c0 time:292862308

**Comment by wangmengyan95** [comment]
Hi @Gelassen, those answers weren't us promising that we worked in china, but at the time we did work in china. In countries where GCM is available, we strongly recommend using GCM rather than PPNS. Therefore, PPNS is no longer included in the default release package.  GCM is much more reliable especially on the newest devices running Android M.  However we understand your use case, and suggest you get PPNS JAR from this [older release](https://www.parse.com/downloads/android/Parse/1.10.2).  Please keep in mind that PPNS connectivity in China is dependent on Chinese Firewall, so we cannot promise that clients in China will always have connectivity to Parse in the future, even if it seems to work in the past.  Our backend service is currently hosted in the US, and we have no plans to host a backend service inside China.

**Comment by Gelassen** [comment]
Hi @wangmengyan95, thank you for clarifying this.
